### PR TITLE
tokenLabel support and EC fixes for Z platform

### PR DESCRIPTION
**Pull request by BobMDu**
- tokenLabel keyword support
- APAR IJ09107
- Point does not match field size Exception for EC curve

Signed-off-by: Bob Du <bob.du@ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
I believe this is the forward port of https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/598

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java [hunk 1]
```diff
@@ -468,6 +483,14 @@ protected void engineInitVerify(PublicKey publicKey)
         if (publicKey == null) {
             throw new InvalidKeyException("Key must not be null");
         }
+
+        // Check if the public key is EC public key to keep its key size
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java:487]: Please finish each sentence with a period.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java [hunk 2]
```diff
@@ -468,6 +483,14 @@ protected void engineInitVerify(PublicKey publicKey)
...
+        if (publicKey instanceof ECPublicKey) {
+            ecKeySize = ((ECPublicKey)publicKey).getParams().getCurve().getField().getFieldSize();
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java:489]: Please use pattern matching:
```java
        if (publicKey instanceof ECPublicKey ecPublicKey) {
            ecKeySize = ecPublicKey.getParams().getCurve().getField().getFieldSize();
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 3]
```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
                 System.out.println(p11Info);
             }
 
-            if ((slotID < 0) || showInfo) {
+            if ((slotID < 0) || showInfo|| (tokenLabel != null)) {
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:389]: Formatting: space both sides of each binary operator (`||`).

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 4]
```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
...
-                    System.out.println("Slots with tokens: " + toString(slots));
+                    if (isZ) {
+                        System.out.println("Slots[slotID, tokenName]:");
+                        for (int i = 0; i < slots.length; i++) {
+                            System.out.println("[" + i + ", " + new String(p11.C_GetTokenInfo(slots[i]).label).trim() + "]");
+                        }
+                    } else {
+                        System.out.println("All slots: " + toString(slots));
+                        long[] slotsWithTokens = p11.C_GetSlotList(true);
+                        System.out.println("Slots with tokens: " + toString(slotsWithTokens));
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:400]: Please minimize the change relative to upstream, i.e. don't introduce a new local variable:
```java
                    slots = p11.C_GetSlotList(true);
                    System.out.println("Slots with tokens: " + toString(slots));
```
The use of `slots` below for Z will be unaffected.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 5]
```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
...
+                        String label = null;
+                        try {
+                            label = new String(p11.C_GetTokenInfo(slots[i]).label).trim();
+                            if (tokenLabel.equalsIgnoreCase(label)) {
+                                slotID = -1;
+                                slotListIndex = i;
+                                found = true;
+                                break;
+                            }
+                        } catch (Exception ex) {
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:416]: What exceptions are you anticipating? Can this be narrowed to catch fewer, more specific exceptions?

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Please also make the summary of your commit message and this pull request less general.
"Updates and fixes for z platform" doesn't suggest the narrow scope of the actual changes.

**Comment by BobMDu** [comment]
@keithc-ca I've made changes to address all the items you pointed out. Can you please review again when you have time? Thanks.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java [hunk 6]
```diff
@@ -500,9 +513,13 @@ private void parse() throws IOException {
...
-                throw new ConfigurationException
-                        ("Unknown keyword '" + st.sval + "', line " +
-                        st.lineno());
+                if ("tokenLabel".equalsIgnoreCase(st.sval)) {
+                    tokenLabel = parseStringEntry("tokenLabel");
+                } else {
+                    throw new ConfigurationException
+                            ("Unknown keyword '" + st.sval + "', line " +
+                                    st.lineno());
+                }
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java:522]: The braces I suggested around this code are required; it doesn't compile without them.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java [hunk 7]
```diff
@@ -162,6 +168,15 @@ final class P11Signature extends SignatureSpi {
     // than 1024 bits", but this is a little arbitrary
     private static final int RAW_ECDSA_MAX = 128;
 
+    // Check if running on z platform
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Signature.java:171]: Please finish each sentence with a period; here, and on lines 178 and 491.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 8]
```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
...
+                            label = new String(p11.C_GetTokenInfo(slots[i]).label).trim();
+                            if (tokenLabel.equalsIgnoreCase(label)) {
+                                slotID = -1;
+                                slotListIndex = i;
+                                found = true;
+                                break;
+                            }
+                        } catch (PKCS11Exception ex) {
+                            label = null;
+                        }
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:418]: Please move `label` into the `try` block; it's not needed elsewhere:
```java
                        try {
                            String label = new String(p11.C_GetTokenInfo(slots[i]).label).trim();
                            if (tokenLabel.equalsIgnoreCase(label)) {
                                slotID = -1;
                                slotListIndex = i;
                                found = true;
                                break;
                            }
                        } catch (PKCS11Exception ex) {
                            // ignore
                        }
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 9]
```diff
@@ -377,12 +386,41 @@ private static <T> T checkNull(T obj) {
...
+                    }
+                    if (!found) {
+                        throw new IOException("Invalid Token Label : "
+                                + tokenLabel);
+                    }
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:423]: I'm not a big fan of redundant variables. The `for` loop above could begin like this:
```java
                    for (int i = 0;; i++) {
                        if (i >= slots.length) {
                            throw new IOException("Invalid Token Label : "
                                    + tokenLabel);
                        }
```
(but I won't insist).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Built locally (with patches to openj9).

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java [hunk 10]
```diff
@@ -500,9 +513,15 @@ private void parse() throws IOException {
             case "nssOptimizeSpace"->
                 nssOptimizeSpace = parseBooleanEntry(st.sval);
             default->
-                throw new ConfigurationException
-                        ("Unknown keyword '" + st.sval + "', line " +
-                        st.lineno());
+                {
+                    if ("tokenLabel".equalsIgnoreCase(st.sval)) {
+                        tokenLabel = parseStringEntry("tokenLabel");
```
[the remark this answers is not in the document]
Comment (BobMDu) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java:518]: @keithc-ca https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/548

**Comment by pshipton** [comment]
It seems we could have a release note / What's New in the doc for the "tokenLabel keyword support". @BobMDu if you agree pls open an issue at https://github.com/eclipse-openj9/openj9-docs

**Pull request merged**